docs/posts/duckdb-for-rag/index.qmd

```diff
@@ -0,0 +1,510 @@
+between arrays of numbers. Alternatively, a custom user-defined function (UDF)
+can be used for the similarity metric.
+:::
+
+[DuckDB is the default backend for Ibis](../why-duckdb/index.qmd) and makes
+rapid prototyping locally very easy. You can deploy it to
+[MotherDuck](https://motherduck.com) just by changing the connection string, or
+swap out DuckDB for another Ibis backend.
+
+## What is RAG?
+
+Retrieval-augmented generation (RAG) is a technique for language model systems
+to retrieve relevant information from a knowledge base, augment the model's
+input with that knowledge, and then generate a response. It serves to provide
+useful context given language models' context length limitations and increase
+the quality of the response.
+
+::: {.callout-note title="Is RAG dead?"}
+With increasing language model context lengths, is RAG dead?
+
+**No**. Though as we'll discuss later, it's unclear if traditional vector
+similarity search RAG is the best approach. Regardless, just because a language
+model can use the entire content of a textbook doesn't mean you'll get better
+performance (accuracy or efficiency) than first retrieving the relevant subsets
+of the textbook.
+:::
+
+Traditional RAG embeds text into vectors of numbers and then computes a
+similarity metric between input text and the knowledge base to determine the
+most relevant context.
+
+```{mermaid}
+graph LR
+    style T fill:#1E90FF,stroke:#1E90FF,color:#ffffff
+    style N fill:#1E90FF,stroke:#1E90FF,color:#ffffff
+    style C fill:#1E90FF,stroke:#1E90FF,color:#ffffff
+
+    T("text") --> N("numbers (embedding vector)")
+    N --> C("compute similarity metric")
+```
+
+Let's jump into the code!
+
+## RAG over Ibis documentation
+
+We'll use the Ibis documentation as our knowledge base.
+
+### Setup code
+
+First we need to install and import the necessary packages:
+
+```bash
+pip install 'ibis-framework[duckdb]' openai python-dotenv
+```
+
+```{python}
+import os  # <1>
+import ibis
+import shutil
+import subprocess
+
+from openai import OpenAI
+from dotenv import load_dotenv
+from pathlib import Path  # <1>
+```
+
+1. Import the libraries we'll use.
+
+We need to set the `OPENAI_API_KEY` environment variable to our OpenAI API key.
+Do this by creating a `.env` file in the same directory as the code and adding
+the following line:
+
+```txt
+OPENAI_API_KEY=<your-api-key>  # <1>
+```
+
+1. Replace with your OpenAI API key.
+
+Then load the environment variables:
+
+```{python}
+load_dotenv()  # <1>
+```
+
+1. Load environment variables from the `.env` file.
+
+Now we're ready to make OpenAI calls.
+
+Let's also setup Ibis for interactive use:
+
+```{python}
+ibis.options.interactive = True  # <1>
+```
+
+1. Turn on Ibis interactive mode.
+
+### Getting the data into DuckDB
+
+We'll use a function to clone a given repository:
+
+```{python}
+def clone_repo(repo_url):  # <1>
+    subprocess.run(["git", "clone", "--depth", "1", repo_url], check=False)  # <1>
```

Comment:
Can't you clone to `tempfile.gettempdir()` instead of having to delete using `rmtree`?

Reply:
why do you think I would know these things. ima PM

Reply:
Why clone at all even? Couldn't you refer to your local editable module?

```python
Path(ibis.__file__).parents[1].joinpath("docs").rglob("*.qmd")
```

Reply:
yes but wanted to make it easy for someone to copy and run w/ their own repo (just change the repo url/name)

Reply:
I'll change to that though and just add a tip about changing the whatever to your own docs